### PR TITLE
kvutils: Add structured error data to the error metadata [KVL-1032]

### PR DESCRIPTION
**Pull request by hubert-da**
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala [hunk 1]
```diff
@@ -590,15 +616,17 @@ private[state] object Conversions {
           buildStatus(
             Code.INVALID_ARGUMENT,
             s"Party not known on ledger: Submitting party '${rejection.getSubmitterParty}' not known",
+            Map("submitter_party" -> rejection.getSubmitterParty),
           )
         )
       case DamlTransactionRejectionEntry.ReasonCase.PARTIES_NOT_KNOWN_ON_LEDGER =>
         val rejection = entry.getPartiesNotKnownOnLedger
+        val parties = rejection.getPartiesList.asScala.mkString("[", ",", "]")
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:624]: I think we might want to JSON-encode this so strings are correctly quoted. See `JsonStringSerializer` in _contextualized-logging_ for an example of how to do this.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala [hunk 2]
```diff
@@ -567,6 +578,7 @@ private[state] object Conversions {
           buildStatus(
             Code.ABORTED,
             s"Inconsistent: Missing input state for key ${rejection.getKey.toString}",
+            Map("key" -> rejection.getKey.toString),
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:581]: What does `toString` do here?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala [hunk 3]
```diff
@@ -188,39 +221,76 @@ class ConversionsSpec extends AnyWordSpec with Matchers with OptionValues {
         "handle with expected status codes" in {
           forAll(
             Table(
-              ("rejection builder", "code"),
+              ("rejection builder", "code", "expectedAdditionalDetails"),
               (
                 (builder: DamlTransactionRejectionEntry.Builder) =>
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala:226]: You can get rid of this boilerplate by declaring an explicit type for the `Table`, i.e., `Table[DamlTransactionRejectionEntry.Builder => DamlTransactionRejectionEntry.Builder, Status.Code]`. Then, the builders can be constructed simply as e.g. `_.setInconsistent(...)`.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala [hunk 4]
```diff
@@ -105,68 +107,97 @@ class ConversionsSpec extends AnyWordSpec with Matchers with OptionValues {
...
             ),
             (
               Rejection.PartiesNotKnownOnLedger(Seq.empty),
               Code.INVALID_ARGUMENT,
+              Map.empty[String, String],
+            ),
+            (
+              Rejection.MissingInputState(partyStateKey("party")),
+              Code.ABORTED,
+              Map("key" -> "party: \"party\"\n"),
```
[the remark this answers is not in the document]
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala:181]: Decided to leave it, as this is what the `TextFormat` is able to parse.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala (hunk 2 above)
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/Conversions.scala:583]: As discussed offline, it uses `TextFormat.printer` underneath.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala (hunk 3 above)
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala]: Great advice, thanks!

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala [hunk 5]
```diff
@@ -105,122 +109,226 @@ class ConversionsSpec extends AnyWordSpec with Matchers with OptionValues {
       "convert rejection to proto models and back to expected grpc code" in {
         forAll(
           Table(
-            ("rejection", "expected code"),
+            ("rejection", "expected code", "expected additional details"),
             (
               Rejection.ValidationFailure(Error.Package(Error.Package.Internal("ERROR", "ERROR"))),
               Code.INVALID_ARGUMENT,
+              Map.empty[String, String],
```
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala:116]: You may be able to remove `[String, String]` from this line and every other if you add types to this `Table` too.

**Review by SamirTalwar-DA — APPROVED** [review]

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala (hunk 5 above)
Comment (hubert-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/ConversionsSpec.scala:116]: Yes, you are right, forgot about this.

**Pull request merged**